In [1]:
!find /kaggle/input -name "attention_baseline.cu"

/kaggle/input/datasets/debasishmohanty8765/gpu-attention-performance-baseline/attention_baseline.cu


In [2]:
!nvcc -O3 -arch=sm_75 \
    /kaggle/input/datasets/debasishmohanty8765/gpu-attention-performance-baseline/attention_baseline.cu \
    -o /kaggle/working/attention_baseline

In [3]:
! /kaggle/working/attention_baseline

=== GPU Attention Baseline V0 ===
Sequence Length : 512
Heads           : 12
Head Dimension  : 64
Precision       : FP16
Warmup          : 20
Iterations      : 100


=== Benchmark ===
Total Time : 432.185 ms
Avg Time   : 4.32185 ms

=== Validation ===
Max Error  : 5.37187e-06
Validation : PASSED


In [4]:
!which nsys
!which ncu

/usr/local/cuda/bin/ncu


In [5]:
!ncu --version

NVIDIA (R) Nsight Compute Command Line Profiler
Copyright (c) 2018-2025 NVIDIA Corporation
Version 2025.1.1.0 (build 35528883) (public-release)


In [6]:
!ncu --set full \
    /kaggle/working/attention_baseline

=== GPU Attention Baseline V0 ===
Sequence Length : 512
Heads           : 12
Head Dimension  : 64
Precision       : FP16
Warmup          : 20
Iterations      : 100

==PROF== Connected to process 211 (/kaggle/working/attention_baseline)
==ERROR== ERR_NVGPUCTRPERM - The user does not have permission to access NVIDIA GPU Performance Counters on the target device 0. For instructions on enabling permissions and to get more information see https://developer.nvidia.com/ERR_NVGPUCTRPERM

=== Benchmark ===
Total Time : 486.879 ms
Avg Time   : 4.86879 ms

=== Validation ===
Max Error  : 5.37187e-06
Validation : PASSED
==PROF== Disconnected from process 211


In [7]:
!nvcc -O3 -arch=sm_75 \
/kaggle/input/datasets/debasishmohanty8765/gpu-attention-performance-baseline/attention_baseline.cu \
-o /kaggle/working/attention_baseline

In [8]:
!/kaggle/working/attention_baseline

=== GPU Attention Baseline V0 ===
Sequence Length : 512
Heads           : 12
Head Dimension  : 64
Precision       : FP16
Warmup          : 20
Iterations      : 100


=== Benchmark ===
Total Time : 470.465 ms
Avg Time   : 4.70465 ms

=== Validation ===
Max Error  : 5.37187e-06
Validation : PASSED


In [11]:
!find /kaggle/input -name "attention_baseline.cu"

/kaggle/input/datasets/debasishmohanty8765/gpu-attention-performance-baseline/attention_baseline.cu


In [12]:
!nvcc -O3 -arch=sm_75 \
/kaggle/input/datasets/debasishmohanty8765/gpu-attention-performance-baseline/attention_baseline.cu \
-o /kaggle/working/attention_baseline

In [13]:
!/kaggle/working/attention_baseline

=== GPU Attention Baseline V0 ===
Sequence Length : 512
Heads           : 12
Head Dimension  : 64
Precision       : FP16
Warmup          : 20
Iterations      : 100


=== Kernel Timing ===
QK Time      : 3.55002 ms
Softmax Time : 0.139611 ms
PV Time      : 0.684278 ms

=== Benchmark ===
Total Time : 472.884 ms
Avg Time   : 4.72884 ms

=== Validation ===
Max Error  : 5.37187e-06
Validation : PASSED


## 🔹 Experiment 0 — Baseline Attention

### Configuration

| Parameter       |           Value |
| --------------- | --------------: |
| GPU             | NVIDIA Tesla T4 |
| Sequence Length |             512 |
| Heads           |              12 |
| Head Dimension  |              64 |
| Precision       |            FP16 |
| Warmup          |              20 |
| Iterations      |             100 |

### Benchmark Result

| Metric       |          Result |
| ------------ | --------------: |
| Total Time   |  **472.884 ms** |
| Average Time |  **4.72884 ms** |
| QK Time      |  **3.55002 ms** |
| Softmax Time | **0.139611 ms** |
| PV Time      | **0.684278 ms** |

### Validation

* Max Error: **5.37187e-06**
* Validation: **PASSED** ✅

### Observation

The baseline attention implementation takes **4.72884 ms** per iteration.

The **QK kernel is the dominant component**, taking approximately **3.55 ms**, while Softmax and PV take significantly less time.

This baseline will be used as the reference for subsequent optimizations.


In [14]:
!find /kaggle/input -name "attention_qk_shared_v1.cu"

/kaggle/input/datasets/debasishmohanty8765/gpu-attention-performance-v1/attention_qk_shared_v1.cu


In [15]:
!nvcc -O3 -arch=sm_75 \
/kaggle/input/datasets/debasishmohanty8765/gpu-attention-performance-v1/attention_qk_shared_v1.cu \
-o /kaggle/working/attention_qk_shared_v1

In [16]:
!/kaggle/working/attention_qk_shared_v1

=== GPU Attention Baseline V0 ===
Sequence Length : 512
Heads           : 12
Head Dimension  : 64
Precision       : FP16
Warmup          : 20
Iterations      : 100


=== Kernel Timing ===
QK Time      : 1.73496 ms
Softmax Time : 0.139923 ms
PV Time      : 0.680815 ms

=== Benchmark ===
Total Time : 302.066 ms
Avg Time   : 3.02066 ms

=== Validation ===
Max Error  : 5.37187e-06
Validation : PASSED


## 🔹 Experiment 1 — QK Shared Memory Optimization

### Optimization

The QK kernel was modified to use **shared memory tiling** for Q and K.

The rest of the attention pipeline remains unchanged.

### Configuration

| Parameter       |           Value |
| --------------- | --------------: |
| GPU             | NVIDIA Tesla T4 |
| Sequence Length |             512 |
| Heads           |              12 |
| Head Dimension  |              64 |
| Precision       |            FP16 |
| Warmup          |              20 |
| Iterations      |             100 |

### Benchmark Result

| Metric       | Baseline V0 |             V1 |      Change |
| ------------ | ----------: | -------------: | ----------: |
| QK Time      |  3.55002 ms | **1.73496 ms** | **-51.13%** |
| Softmax Time | 0.139611 ms |    0.139923 ms |       ~same |
| PV Time      | 0.684278 ms |    0.680815 ms |       ~same |
| Average Time |  4.72884 ms | **3.02066 ms** | **-36.10%** |

### Speedup

**End-to-end speedup: 1.565×**

### Validation

* Max Error: **5.37187e-06**
* Validation: **PASSED** ✅

### Observation

The shared-memory QK optimization reduced QK kernel time from **3.55002 ms → 1.73496 ms**, giving approximately **51% lower QK latency**.

Since QK was the dominant part of the baseline, this reduced overall attention latency from **4.72884 ms → 3.02066 ms**.

The numerical result remains unchanged within the validation threshold.


In [1]:
!find /kaggle/input -name "attention_pv_shared_v2.cu"

/kaggle/input/datasets/debasishmohanty8765/attention-pv-shared-v2/attention_pv_shared_v2.cu


In [2]:
!nvcc -O3 -arch=sm_75 \
/kaggle/input/datasets/debasishmohanty8765/attention-pv-shared-v2/attention_pv_shared_v2.cu \
-o /kaggle/working/attention_pv_shared_v2

In [3]:
!/kaggle/working/attention_pv_shared_v2

=== GPU Attention V2 (QK + PV shared memory) ===
Sequence Length : 512
Heads           : 12
Head Dimension  : 64
Precision       : FP16
Warmup          : 20
Iterations      : 100


=== Kernel Timing ===
QK Time      : 1.7511 ms
Softmax Time : 0.140719 ms
PV Time      : 0.429425 ms

=== Benchmark ===
Total Time : 281.819 ms
Avg Time   : 2.81819 ms

=== Validation ===
Max Error  : 5.37187e-06
Validation : PASSED


## 🔹 Experiment 2 — QK + PV Shared Memory Optimization

### Optimization

V2 extends the previous optimization by applying **shared memory tiling to the PV kernel**.

The QK kernel continues to use the shared-memory optimization introduced in V1.

The Softmax kernel remains unchanged.

### Configuration

| Parameter       |           Value |
| --------------- | --------------: |
| GPU             | NVIDIA Tesla T4 |
| Sequence Length |             512 |
| Heads           |              12 |
| Head Dimension  |              64 |
| Precision       |            FP16 |
| Warmup          |              20 |
| Iterations      |             100 |

### Benchmark Result

| Metric       | Baseline V0 |          V1 |              V2 |    V2 vs V0 |
| ------------ | ----------: | ----------: | --------------: | ----------: |
| QK Time      |  3.55002 ms |  1.73496 ms |  **1.75110 ms** |     -50.67% |
| Softmax Time | 0.139611 ms | 0.139923 ms | **0.140719 ms** |      +0.79% |
| PV Time      | 0.684278 ms | 0.680815 ms | **0.429425 ms** | **-37.24%** |
| Average Time |  4.72884 ms |  3.02066 ms |  **2.81819 ms** | **-40.40%** |

### Speedup

**End-to-end speedup vs V0: 1.678×**

V2 reduces average attention latency from:

**4.72884 ms → 2.81819 ms**

This is approximately a **40.40% reduction in latency**.

### PV Improvement

The main improvement in V2 comes from the PV kernel:

**0.680815 ms → 0.429425 ms**

This represents approximately a **37.24% reduction in PV kernel latency**.

The QK kernel remains approximately the same as V1:

**1.73496 ms → 1.75110 ms**

Therefore, the V2 improvement should primarily be attributed to the **PV shared-memory optimization**, not QK.

### Validation

* Max Error: **5.37187e-06**
* Validation: **PASSED** ✅

The numerical result remains within the validation threshold, confirming that the PV optimization preserves correctness.

### Observation

V2 further reduces attention latency by optimizing the PV kernel with shared memory.

The dominant QK bottleneck was already reduced significantly in V1. V2 targets another expensive memory-access pattern in the PV operation.

Overall, the attention workload improved from **4.72884 ms in V0 to 2.81819 ms in V2**, achieving approximately **1.68× end-to-end speedup** while maintaining numerical correctness.


In [4]:
import pandas as pd
import matplotlib.pyplot as plt
from pathlib import Path
import zipfile

out = Path("/mnt/data/attention_benchmark_results")
out.mkdir(parents=True, exist_ok=True)

# -----------------------------
# Benchmark data
# -----------------------------
df = pd.DataFrame([
    ["V0", 3.55002, 0.139611, 0.684278, 4.72884],
    ["V1", 1.73496, 0.139923, 0.680815, 3.02066],
    ["V2", 1.75110, 0.140719, 0.429425, 2.81819],
], columns=["Version", "QK_ms", "Softmax_ms", "PV_ms", "Avg_ms"])

baseline = df.loc[0, "Avg_ms"]
df["Speedup_vs_V0"] = baseline / df["Avg_ms"]
df["Latency_Reduction_vs_V0_pct"] = (1 - df["Avg_ms"] / baseline) * 100
df["QK_Reduction_vs_V0_pct"] = (1 - df["QK_ms"] / df.loc[0, "QK_ms"]) * 100
df["PV_Reduction_vs_V0_pct"] = (1 - df["PV_ms"] / df.loc[0, "PV_ms"]) * 100

# -----------------------------
# Validation
# -----------------------------
validation = pd.DataFrame([
    ["V0", 5.37187e-06, "PASSED"],
    ["V1", 5.37187e-06, "PASSED"],
    ["V2", 5.37187e-06, "PASSED"],
], columns=["Version", "Max_Error", "Validation"])

# -----------------------------
# Configuration
# -----------------------------
config = pd.DataFrame([
    ["GPU", "NVIDIA Tesla T4"],
    ["Sequence Length", 512],
    ["Heads", 12],
    ["Head Dimension", 64],
    ["Precision", "FP16"],
    ["Warmup", 20],
    ["Iterations", 100],
], columns=["Parameter", "Value"])

# -----------------------------
# Experiment summary
# -----------------------------
experiments = pd.DataFrame([
    ["V0", "Baseline Attention", "QK + Softmax + PV baseline", "4.72884 ms", "1.000×"],
    ["V1", "QK Shared Memory", "Shared-memory tiling for QK", "3.02066 ms", "1.565×"],
    ["V2", "QK + PV Shared Memory", "Shared-memory tiling for QK and PV", "2.81819 ms", "1.678×"],
], columns=["Version", "Experiment", "Optimization", "Average_Latency", "Speedup_vs_V0"])

# -----------------------------
# Save CSVs
# -----------------------------
df.to_csv(out / "attention_benchmark_results.csv", index=False)
df[["Version", "QK_ms", "Softmax_ms", "PV_ms"]].to_csv(
    out / "attention_kernel_timings.csv", index=False
)
validation.to_csv(out / "attention_validation.csv", index=False)
config.to_csv(out / "attention_configuration.csv", index=False)
df.to_csv(out / "attention_comparison.csv", index=False)
experiments.to_csv(out / "attention_experiments.csv", index=False)

# -----------------------------
# Plot helper
# -----------------------------
def save_bar(filename, x, y, ylabel, title, labels=None):
    plt.figure(figsize=(8, 5))
    plt.bar(x, y)
    plt.ylabel(ylabel)
    plt.xlabel("Attention version")
    plt.title(title)
    if labels is not None:
        for i, v in enumerate(y):
            plt.text(i, v, labels[i], ha="center", va="bottom")
    plt.tight_layout()
    plt.savefig(out / filename, dpi=200)
    plt.close()

# 1. End-to-end latency
save_bar(
    "01_end_to_end_latency.png",
    df["Version"],
    df["Avg_ms"],
    "Average latency (ms)",
    "Attention End-to-End Latency",
    [f"{v:.3f} ms" for v in df["Avg_ms"]]
)

# 2. Kernel timing comparison
x = list(range(len(df)))
width = 0.25
plt.figure(figsize=(9, 5))
plt.bar([i-width for i in x], df["QK_ms"], width, label="QK")
plt.bar(x, df["Softmax_ms"], width, label="Softmax")
plt.bar([i+width for i in x], df["PV_ms"], width, label="PV")
plt.xticks(x, df["Version"])
plt.ylabel("Kernel time (ms)")
plt.xlabel("Attention version")
plt.title("Kernel Timing Comparison")
plt.legend()
plt.tight_layout()
plt.savefig(out / "02_kernel_timing_comparison.png", dpi=200)
plt.close()

# 3. Speedup
save_bar(
    "03_speedup_vs_v0.png",
    df["Version"],
    df["Speedup_vs_V0"],
    "Speedup (×)",
    "End-to-End Speedup vs V0",
    [f"{v:.3f}×" for v in df["Speedup_vs_V0"]]
)

# 4. Latency reduction
save_bar(
    "04_latency_reduction_vs_v0.png",
    df["Version"],
    df["Latency_Reduction_vs_V0_pct"],
    "Latency reduction (%)",
    "Latency Reduction vs V0",
    [f"{v:.2f}%" for v in df["Latency_Reduction_vs_V0_pct"]]
)

# 5. Stacked kernel breakdown
plt.figure(figsize=(9, 5))
plt.bar(df["Version"], df["QK_ms"], label="QK")
plt.bar(df["Version"], df["Softmax_ms"], bottom=df["QK_ms"], label="Softmax")
plt.bar(
    df["Version"],
    df["PV_ms"],
    bottom=df["QK_ms"] + df["Softmax_ms"],
    label="PV"
)
plt.ylabel("Kernel time (ms)")
plt.xlabel("Attention version")
plt.title("Attention Kernel Breakdown")
plt.legend()
plt.tight_layout()
plt.savefig(out / "05_kernel_breakdown_stacked.png", dpi=200)
plt.close()

# 6. Kernel improvement
plt.figure(figsize=(9, 5))
plt.plot(
    df["Version"],
    df["QK_Reduction_vs_V0_pct"],
    marker="o",
    label="QK reduction"
)
plt.plot(
    df["Version"],
    df["PV_Reduction_vs_V0_pct"],
    marker="o",
    label="PV reduction"
)
plt.ylabel("Reduction vs V0 (%)")
plt.xlabel("Attention version")
plt.title("Kernel Improvement vs Baseline")
plt.legend()
plt.grid(True, alpha=0.25)
plt.tight_layout()
plt.savefig(out / "06_kernel_improvement.png", dpi=200)
plt.close()

# 7. Speedup + latency reduction
fig, ax1 = plt.subplots(figsize=(9, 5))
ax1.bar(df["Version"], df["Speedup_vs_V0"])
ax1.set_xlabel("Attention version")
ax1.set_ylabel("Speedup (×)")
ax1.set_title("Performance Improvement Across Versions")
for i, v in enumerate(df["Speedup_vs_V0"]):
    ax1.text(i, v, f"{v:.3f}×", ha="center", va="bottom")
plt.tight_layout()
plt.savefig(out / "07_performance_improvement.png", dpi=200)
plt.close()

# -----------------------------
# Markdown report
# -----------------------------
report = f"""# GPU Attention Performance — V0 / V1 / V2

## Configuration

| Parameter | Value |
|---|---|
| GPU | NVIDIA Tesla T4 |
| Sequence Length | 512 |
| Heads | 12 |
| Head Dimension | 64 |
| Precision | FP16 |
| Warmup | 20 |
| Iterations | 100 |

## Benchmark

| Version | QK (ms) | Softmax (ms) | PV (ms) | Average (ms) | Speedup |
|---|---:|---:|---:|---:|---:|
| V0 | 3.55002 | 0.139611 | 0.684278 | 4.72884 | 1.000× |
| V1 | 1.73496 | 0.139923 | 0.680815 | 3.02066 | 1.565× |
| V2 | 1.75110 | 0.140719 | 0.429425 | 2.81819 | 1.678× |

## Validation

All versions passed validation with Max Error = 5.37187e-06.

## Key Results

- V1 QK reduction: {df.loc[1, "QK_Reduction_vs_V0_pct"]:.2f}%
- V2 PV reduction: {df.loc[2, "PV_Reduction_vs_V0_pct"]:.2f}%
- V2 end-to-end latency reduction: {df.loc[2, "Latency_Reduction_vs_V0_pct"]:.2f}%
- V2 end-to-end speedup: {df.loc[2, "Speedup_vs_V0"]:.3f}×

## Experiments

### V0 — Baseline
Reference implementation with baseline QK, Softmax and PV kernels.

### V1 — QK Shared Memory
QK uses shared-memory tiling. QK latency improved from 3.55002 ms to 1.73496 ms.

### V2 — QK + PV Shared Memory
PV also uses shared-memory tiling. PV latency improved to 0.429425 ms, reducing end-to-end latency to 2.81819 ms.

## Conclusion

V2 improves the baseline from 4.72884 ms to 2.81819 ms per iteration,
achieving approximately 1.68× end-to-end speedup while maintaining numerical correctness.
"""
(out / "attention_benchmark_report.md").write_text(report)

# -----------------------------
# README
# -----------------------------
readme = """# Attention Benchmark Artifacts

Generated from measured NVIDIA Tesla T4 results for V0, V1 and V2.

## Files

- attention_benchmark_results.csv
- attention_kernel_timings.csv
- attention_validation.csv
- attention_configuration.csv
- attention_comparison.csv
- attention_experiments.csv
- attention_benchmark_report.md
- 01_end_to_end_latency.png
- 02_kernel_timing_comparison.png
- 03_speedup_vs_v0.png
- 04_latency_reduction_vs_v0.png
- 05_kernel_breakdown_stacked.png
- 06_kernel_improvement.png
- 07_performance_improvement.png
"""
(out / "README.md").write_text(readme)

# -----------------------------
# ZIP
# -----------------------------
zip_path = Path("/mnt/data/attention_benchmark_artifacts.zip")
with zipfile.ZipFile(zip_path, "w", zipfile.ZIP_DEFLATED) as z:
    for p in sorted(out.iterdir()):
        z.write(p, arcname=f"attention_benchmark_results/{p.name}")

print(f"Created: {out}")
print(f"ZIP: {zip_path}")
print("\nFiles:")
for p in sorted(out.iterdir()):
    print(" -", p.name)


Created: /mnt/data/attention_benchmark_results
ZIP: /mnt/data/attention_benchmark_artifacts.zip

Files:
 - 01_end_to_end_latency.png
 - 02_kernel_timing_comparison.png
 - 03_speedup_vs_v0.png
 - 04_latency_reduction_vs_v0.png
 - 05_kernel_breakdown_stacked.png
 - 06_kernel_improvement.png
 - 07_performance_improvement.png
 - README.md
 - attention_benchmark_report.md
 - attention_benchmark_results.csv
 - attention_comparison.csv
 - attention_configuration.csv
 - attention_experiments.csv
 - attention_kernel_timings.csv
 - attention_validation.csv


In [5]:
!ls -lh /mnt/data/attention_benchmark_results

total 404K
-rw-r--r-- 1 root root  44K Oct  5 16:15 01_end_to_end_latency.png
-rw-r--r-- 1 root root  46K Oct  5 16:15 02_kernel_timing_comparison.png
-rw-r--r-- 1 root root  48K Oct  5 16:15 03_speedup_vs_v0.png
-rw-r--r-- 1 root root  50K Oct  5 16:15 04_latency_reduction_vs_v0.png
-rw-r--r-- 1 root root  40K Oct  5 16:15 05_kernel_breakdown_stacked.png
-rw-r--r-- 1 root root  85K Oct  5 16:15 06_kernel_improvement.png
-rw-r--r-- 1 root root  52K Oct  5 16:15 07_performance_improvement.png
-rw-r--r-- 1 root root 1.3K Oct  5 16:15 attention_benchmark_report.md
-rw-r--r-- 1 root root  398 Oct  5 16:15 attention_benchmark_results.csv
-rw-r--r-- 1 root root  398 Oct  5 16:15 attention_comparison.csv
-rw-r--r-- 1 root root  123 Oct  5 16:15 attention_configuration.csv
-rw-r--r-- 1 root root  276 Oct  5 16:15 attention_experiments.csv
-rw-r--r-- 1 root root  117 Oct  5 16:15 attention_kernel_timings.csv
-rw-r--r-- 1 root root   95 Oct  5 16:15 attention_validation.csv
-rw-r--r-- 1 root roo

In [6]:
!cp /mnt/data/attention_benchmark_artifacts.zip /kaggle/working/